### ISSUE TITLE
Add `is_outputs_batched` param to `autograd.grad`

### ISSUE THREAD

**Issue by ValerianRey** [pytorch/pytorch#156616]
### 🚀 The feature, motivation and pitch

I want to compute per-sample gradients efficiently, but I can't use the `torch.func` API for this, so the tips given in [this tutorial](https://docs.pytorch.org/tutorials/intermediate/per_sample_grads.html) are not applicable.

<details>
<summary>Details on why I can't use the functional API</summary>

> In my case, I need to compute the per-sample gradients of the loss with respect to the model parameters **separately for each model parameter tensor**. This is because I will compute some stats based on these gradients (namely all of their inner products), so right after these stats are computed, I do not have to store them in memory anymore. Computing them one parameter tensor at a time is thus orders of magnitude more memory-efficient than computing them all at once.
> Since I need to do the entire forward pass before having a "vector" to provide for the vector-Jacobian products, I can't just parallelize a single call to `torch.func.vjp` and to the vjp function that this function returns. I would have to parallelize the call to `torch.func.vjp`, store the batch of output values and the "batch of vjp functions", and then during the backward pass I would have to call this "batch of vjp functions" in parallel. As far as I know, there is no way that `vmap` can return a list of functions, or that `vmap` can parallelize over an existing list of functions.
> So for these combined reasons, I'm stuck with having to use the `autograd` API. If you think I can actually solve my problem with `torch.func`, I would be very happy to hear a solution.
> 
> My current alternative is to do the entire forward pass first, compute my required `grad_outputs` using `autograd.grad`, and then make a vmapped call to `torch.func.vjp` and to the vjp function returned by `torch.func.vjp`, but this comes at the expense of an un-necessary extra forward pass through each layer. It's decent but far from optimal, and it's not very elegant since it makes me have to mix the `autograd` and the `func` APIs.  

</details>

This simply amounts to differentiating a batch of outputs in parallel, with respect to the model parameters, and optionally using a batch of grad_outputs if the outputs are not scalar. 

I saw that there exists a `is_grads_batched` parameter in `torch.autograd.grad`. From what I read in the [PR that added this parameter](https://github.com/pytorch/pytorch/pull/65564), the same result can be equivalently achieved by a vmapped (over the `grad_outputs`) call to `torch.autograd.grad`, so this parameter is mainly here for syntactic sugar or for users that are not aware of `vmap`. However, in my case, a vmapped (over both the `outputs` and the `grad_outputs`) call to `torch.autograd.grad` **does not work**.

I asked for help on the PyTorch forums (see discussion [here](https://discuss.pytorch.org/t/cant-vmap-autograd-grad-over-outputs/220878)) but no one seems to have found a satisfying solution.

I also looked at existing issues about this, and I found this very related issue: https://github.com/pytorch/pytorch/issues/132378. In this issue's case, a `is_outputs_batched` parameter would have elegantly solved the person's problem (although in their case the vmapped call to one of the functional-API functions would also work - requiring a bit of boilerplate code to transform the model into a function).

Here is a short example of how I would expect a `is_outputs_batched` parameter to work:

```python
import torch
from torch import nn

batch_size = 16
model = nn.Linear(3, 1)
input = torch.randn((batch_size, 3))
output = model(input)  # Batch of scalar outputs, of shape [16, 1]

per_sample_grads = torch.autograd.grad(outputs=output, inputs=list(model.parameters()), is_outputs_batched=True)
```
Here, we should obtain for each model parameter `p` a tensor of shape `[batch_size] + p.shape` (i.e. a Jacobian rather than a single gradient).

Note that it should be compatible with `is_grads_batched` in case the outputs are not scalar:
```python
import torch
from torch import nn

batch_size = 16
model = nn.Linear(3, 5)
input = torch.randn((batch_size, 3))
output = model(input)  # Batch of vector outputs, of shape [16, 5]
grad_outputs = torch.ones_like(output)  # Using ones for the sake of the example, but in reality this could come from the differentiation of a later part of the model

per_sample_grads = torch.autograd.grad(outputs=output, inputs=list(model.parameters()), grad_outputs=grad_outputs, is_outputs_batched=True, is_grads_batched=True)
```
Here, again, we should obtain for each model parameter `p` a tensor of shape `[batch_size] + p.shape`.

What do you think? Would it even be possible to add such a feature?

### Alternatives

As far as I know, the only alternative is to do the forward pass and part of the backward pass with the `torch.autograd` API, and doing an extra forward pass and the rest of the backward pass using calls to `torch.func.vjp` vmapped over the batch dimension. This duplicates the computations of the forward pass of each parameterized layer.

### Additional context

With my research partner, we're working on what we call instance-wise risk minimization (see [our paper](https://arxiv.org/pdf/2406.16232v3), Eq. 8). The idea is to decrease (or at least not increase) the loss of all elements of the batch at each training iteration, rather than just decrease the average loss over the batch. For this, we can compute the per-sample gradients (together, they form a Jacobian matrix), and select an update direction with a negative inner-product with all of these gradients. However, we realized that in order to do this, we only need all of the inner product between per-sample gradients (i.e. the gramian of the Jacobian matrix). A much more memory-efficient implementation is thus to compute this gramian as a sum of sub-gramians, which is why we want to compute these per-sample gradients on a parameter-by-parameter basis (see Algorithm 3 of [our paper](https://arxiv.org/pdf/2406.16232v3)). This is why we cannot simply use the solutions of [this tutorial](https://docs.pytorch.org/tutorials/intermediate/per_sample_grads.html) (i.e. using a vmapped call to `torch.func` to get per-sample gradients for the whole model at once).

We have made a library called [TorchJD](https://github.com/TorchJD/torchjd) to reproduce and extend our research, and we're currently implementing the Gramian-based version (https://github.com/TorchJD/torchjd/pull/387), but struggling to make it as optimized as it could. We think that a `is_outputs_batched` parameter would avoid having to do the forward pass twice, which would make our method closer to being usable in practice.

**Comment by albanD** [comment]
Hey!
Isn't this covered by https://dev-discuss.pytorch.org/t/an-efficient-way-to-compute-per-sample-gradients/695 ?

**Comment by soulitzer** [comment]
Yeah ExpandedWeights as Alban mentions sounds like it should apply to this case, although I wonder if the current level of coverage is enough for this case.

Btw if you are okay with writing your forward in a per-sample way, then maybe torch.func can still work for you, because if the problem with torch.func.grad + vmap is that you need the function outputs in order to compute grad_outputs, I wonder if using torch.func.vjp instead of torch.func.grad solves this problem for you?

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.func import functional_call, vjp, vmap

batch_size = 64

inputs = torch.randn((batch_size, 5))
targets = torch.randn((batch_size, 1))

model = nn.Linear(5, 1)

def compute_loss(params, x , t):
    # Compute loss for a single sample
    x = x.unsqueeze(0)
    t = t.unsqueeze(0)
    y = functional_call(model, params, x)
    return F.mse_loss(y, t, reduction="none").squeeze()

# vmap does not support returning non-Tensor as outputs, so we need to
# smuggle the vjp function out of the vmap
vjpfunc = None

def vjp_(inputs, targets):
    global vjpfunc
    out, vjpfunc_ = vjp(compute_loss, {k:v for k, v in model.named_parameters()}, inputs, targets)
    vjpfunc = vjpfunc_
    return out

out = vmap(vjp_)(inputs, targets)
grad = vmap(vjpfunc)(torch.ones(batch_size))
```

@zou3519 for more thoughts

**Comment by ValerianRey** [comment]
> Hey! Isn't this covered by https://dev-discuss.pytorch.org/t/an-efficient-way-to-compute-per-sample-gradients/695 ?

Hey! Thanks for this suggestion, I didn't know about this solution or about all the linked discussions surrounding per-sample gradients. This was very helpful. I managed to make a working implementation of my algorithm using this, but I still have to make two forward passes. To prevent it, I would need this `call_for_per_sample_grads(model)(input)` to be able to make a graph that will allow me to compute both per-sample grads (to be stored in `grad_sample`) and one single grad (to be stored in `grad`). This is quite specific to my use-case, because I need:
1) A single forward pass
2) A backward pass computing per-sample grads, using the graph computed during 1) (and retaining this graph). Compute weights from them.
3) A backward pass computing a single grad, using the loss weighted based on the weights computed in 2), and reusing the graph computed during 1).

It also seemed a bit slower (~20%) than my current solution, so I do not think that this actually covers my use-case.

**Comment by ValerianRey** [comment]
> Btw if you are okay with writing your forward in a per-sample way, then maybe torch.func can still work for you, because if the problem with torch.func.grad + vmap is that you need the function outputs in order to compute grad_outputs, I wonder if using torch.func.vjp instead of torch.func.grad solves this problem for you?

Thanks for your code example! I was already using `torch.func.vjp`, but I didn't know that it was possible to store the vjpfunc in the outside scope of a vmapped function. This is actually a solution to my problem, because I can then turn the forward pass into a composition of vmapped calls to `vjp_` (one for each layer of my Sequential model), and store all of the associated `vjpfunc` in a list.

I can then do both of my backward passes using these stored `vjpfunc`.

This requires quite a lot of boilerplate code, but it works, and it allows me to do what I want with a single forward pass and parallelized backward passes. So far, this seems to be a bit slower than what I had before, but I will try to optimize my code and see how it compares.

I will update you next week about the results.

**Comment by ValerianRey** [comment]
Here is my current (quite optimized) implementation using batched vjp calls:

```python
from queue import LifoQueue
from typing import Any

import torch
from torch import nn, Tensor
from torch.nn import Parameter

batch_size = 64

# Note that this model works correctly with non-batched inputs, a requirement to use the
# functional API
model = nn.Sequential(
    nn.Linear(100, 200),
    nn.ReLU(),
    nn.Linear(200, 300),
    nn.ReLU(),
    nn.Linear(300, 200),
    nn.ReLU(),
    nn.Linear(200, 100),
    nn.ReLU(),
    nn.Linear(100, 1),
)


def make_vjp_call_from_module(module: nn.Module):
    """
    From a module, return a function that takes as input the inputs to the forward function
    of this module, and which is considered as a function of both these inputs and the module's
    parameters, returning both the output and the corresponding vjp function.
    """

    def vjp_from_module(*inputs) -> tuple[Tensor, Any]:
        def functional_model_call(primals: dict[str, Parameter], *inputs) -> Tensor:
            all_state = {**primals, **dict(module.named_buffers())}
            return torch.func.functional_call(module, all_state, *inputs)

        return torch.func.vjp(functional_model_call, dict(module.named_parameters()), *inputs)

    return vjp_from_module


x = torch.randn((batch_size, 100))

vjpfuncs = LifoQueue()
for layer in model:
    # Obtain the next output and store the corresponding vjp_func for later use during the
    # backward pass
    vjp_call = make_vjp_call_from_module(layer)
    x, vjp_func = torch.vmap(vjp_call, out_dims=(0, None))(x)
    vjpfuncs.put(vjp_func)

# For the sake of the example, consider the final value as the vector of losses
losses = x  # shape [batch_size, 1]
grad = torch.ones_like(losses)

gramian = torch.zeros((batch_size, batch_size))

for _ in range(len(model)):
    vjp_function = torch.vmap(vjpfuncs.get())

    # Differentiate w.r.t. both the layer's parameters and the previous activation
    jacobians, grad = vjp_function(grad)
    
    # Accumulate the gramian of the jacobian of each of the layer's parameter into the global gramian
    for jacobian in jacobians.values():
        J = jacobian.reshape((batch_size, -1))
        gramian.addmm_(J, J.T)

# For simplicity, use a very simple rule to obtain the weights (in reality this is more complex)
weights = gramian.sum(dim=0).unsqueeze(1)  # Shape: [batch_size, 1]

# Perform the second backward pass with the obtained weights.
losses.backward(weights)
```
This works. It's still a mix of the functional API and the autograd API (see that the second backward pass is made through a simple `losses.backward(weights)` call. It does avoid the double forward pass problem.

A few notes:
- I didn't actually need to store the `vjp_func` returned by `torch.func.vjp` into a `global` or `nonlocal` variable, because returning it works if I use `None` for the `out_dims` of this output value in my call to `torch.vmap`.
- For simplicity, I always differentiate w.r.t. both the input (the previous activation) and the layer's parameters. In my performance test, I avoided differentiating w.r.t. the input in the very first layer, because we (generally) don't care about derivatives w.r.t. the model input.
- I also tried making the second backward pass by calling again the `vjpfuncs` myself, but it's shorter and similarly efficient to just use autograd for this (`losses.backward(weights)`).

This is, however, still a bit (5-20%) slower than what I used before (a method mixing autograd and vmapped `torch.func.vjp` calls during the first backward pass).

So from my perspective, I would say that I'm still interested in seing an `is_outputs_batched` param in `autograd.grad`. But I also understand that this is maybe a relatively niche use-case, as it would only help people that are interested in per-sample grads and that are unsatisfied with the performance or usability of the `torch.func` API.

Thanks again for all the suggestions btw, they really helped deepen my understanding.
